### PR TITLE
Vertex AI

### PR DESCRIPTION
**Pull request by kuraleta**

### CODE REVIEW

File: langchain4j-vertex/pom.xml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>dev.langchain4j</groupId>
+        <artifactId>langchain4j-parent</artifactId>
+        <version>0.21.0</version>
+        <relativePath>../langchain4j-parent/pom.xml</relativePath>
+    </parent>
+
+    <artifactId>langchain4j-vertex</artifactId>
```
Comment (langchain4j) [inline langchain4j-vertex/pom.xml]: pls rename into langchain4j-vertex-ai

File: langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VertexAiChatModel.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+        this.publisher = publisher;
+        this.modelName = modelName;
+        this.vertexAiParameters = new VertexAiParameters(temperature, maxOutputTokens, topK, topP);
+        this.maxRetries = maxRetries;
+    }
+
+    @Override
+    public AiMessage sendMessages(List<ChatMessage> messages) {
+        try {
+            PredictionServiceSettings predictionServiceSettings = PredictionServiceSettings.newBuilder()
```
Comment (langchain4j) [inline langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VertexAiChatModel.java]: please move PredictionServiceSettings and EndpointName to constructor

File: langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VertexAiEmbeddingModel.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                        throw new RuntimeException(e);
+                    }
+                });
+
+                List<Value> instances = singletonList(instanceValue.build());
+
+                PredictResponse response = withRetry(() -> client.predict(endpointName, instances, ValueConverter.EMPTY_VALUE), maxRetries);
+                Embedding embedding = Embedding.from(extractContent(response));
+
+                return singletonList(embedding);
```
Comment (langchain4j) [inline langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VertexAiEmbeddingModel.java]: it supports multiple embedding instances (text), why does it return only one embedding?

File: langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VertexAiChatModel.java [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+                    .setEndpoint(endpoint)
+                    .build();
+
+            PredictionServiceClient client = PredictionServiceClient.create(predictionServiceSettings);
```
Comment (langchain4j) [inline langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VertexAiChatModel.java]: please do try-with-resources

File: langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VertexAiEmbeddingModel.java [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+            Value.Builder instanceBuilder = Value.newBuilder();
+            for (String text : texts) {
+                JsonFormat.parser().merge(toJson(new VertexAiEmbeddingInstance(text)), instanceBuilder);
+            }
+            List<Value> instances = singletonList(instanceBuilder.build());
+
+            PredictResponse response = withRetry(() -> client.predict(endpointName, instances, EMPTY_VALUE), maxRetries);
+
+            Embedding embedding = Embedding.from(toVector(response));
```
Comment (langchain4j) [inline langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VertexAiEmbeddingModel.java]: should extract multiple vectors

File: langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VextexAiLanguageModel.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        this.modelName = ensureNotBlank(modelName, "modelName");
+        this.vertexAiParameters = new VertexAiParameters(temperature, maxOutputTokens, topK, topP);
+        this.maxRetries = maxRetries;
+    }
+
+    @Override
+    public String process(String text) {
+
+        try {
+            PredictionServiceSettings predictionServiceSettings = PredictionServiceSettings.newBuilder()
```
Comment (langchain4j) [inline langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VextexAiLanguageModel.java]: pls move to ctor

File: langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VextexAiLanguageModel.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+                    .setEndpoint(endpoint)
+                    .build();
+
+            PredictionServiceClient client = PredictionServiceClient.create(predictionServiceSettings);
```
Comment (langchain4j) [inline langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VextexAiLanguageModel.java]: try with resources

File: langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VextexAiLanguageModel.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+            final EndpointName endpointName = EndpointName.ofProjectLocationPublisherModelName(project, location, publisher, modelName);
```
Comment (langchain4j) [inline langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VextexAiLanguageModel.java]: pls move to ctor

File: langchain4j-vertex/src/test/java/dev/langchain4j/model/vertex/VertexAiChatModelIT.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class VertexAiChatModelIT {
+
+    @Test
+    @Disabled("To run this test, you must have provide your own endpoint, project and location")
+    void testChatModel() {
+        VertexAiChatModel vertexAiChatModel = new VertexAiChatModel(
```
Comment (langchain4j) [inline langchain4j-vertex/src/test/java/dev/langchain4j/model/vertex/VertexAiChatModelIT.java]: pls use builder, not clear what all those parameters are

File: langchain4j-vertex/src/test/java/dev/langchain4j/model/vertex/VextexAiLanguageModelIT.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class VextexAiLanguageModelIT {
+
+    @Test
+    @Disabled("To run this test, you must have provide your own endpoint, project and location")
+    void testLanguageModel() {
+        VextexAiLanguageModel vextexAiLanguageModel = new VextexAiLanguageModel(
```
Comment (langchain4j) [inline langchain4j-vertex/src/test/java/dev/langchain4j/model/vertex/VextexAiLanguageModelIT.java]: no need to test it via ctor, logic should be the same

**Review by ghost — COMMENTED** [review]
good job!

File: langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VextexAiLanguageModel.java (hunk 8 above)
Comment (kuraleta) [reply langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VextexAiLanguageModel.java]: done

File: langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VextexAiLanguageModel.java (hunk 6 above)
Comment (kuraleta) [reply langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VextexAiLanguageModel.java]: done

File: langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VertexAiChatModel.java (hunk 2 above)
Comment (kuraleta) [reply langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VertexAiChatModel.java]: done

File: langchain4j-vertex/pom.xml (hunk 1 above)
Comment (kuraleta) [reply langchain4j-vertex/pom.xml]: done

File: langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VertexAiChatModel.java (hunk 4 above)
Comment (kuraleta) [reply langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VertexAiChatModel.java]: done

File: langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VextexAiLanguageModel.java (hunk 7 above)
Comment (kuraleta) [reply langchain4j-vertex/src/main/java/dev/langchain4j/model/vertex/VextexAiLanguageModel.java]: done

File: langchain4j-vertex/src/test/java/dev/langchain4j/model/vertex/VextexAiLanguageModelIT.java (hunk 10 above)
Comment (kuraleta) [reply langchain4j-vertex/src/test/java/dev/langchain4j/model/vertex/VextexAiLanguageModelIT.java]: ok, deleted

File: langchain4j-vertex/src/test/java/dev/langchain4j/model/vertex/VertexAiChatModelIT.java (hunk 9 above)
Comment (kuraleta) [reply langchain4j-vertex/src/test/java/dev/langchain4j/model/vertex/VertexAiChatModelIT.java]: ok, changed

**Review by ghost — DISMISSED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**